pkg/authorization/admission/clusterquota/controller.go

```diff
@@ -0,0 +1,528 @@
+package clusterquota
+
+import (
+	"fmt"
+	"sort"
+	"sync"
+	"time"
+
+	"k8s.io/kubernetes/pkg/admission"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/meta"
+	"k8s.io/kubernetes/pkg/api/resource"
+	"k8s.io/kubernetes/pkg/quota"
+	utilruntime "k8s.io/kubernetes/pkg/util/runtime"
+	"k8s.io/kubernetes/pkg/util/sets"
+	"k8s.io/kubernetes/pkg/util/wait"
+	"k8s.io/kubernetes/pkg/util/workqueue"
+
+	quotaapi "github.com/openshift/origin/pkg/authorization/api"
+	oclient "github.com/openshift/origin/pkg/client"
+)
+
+type quotaEvaluator struct {
+	client oclient.Interface
+
+	// registry that knows how to measure usage for objects
+	registry quota.Registry
+
+	// TODO these are used together to bucket items by namespace and then batch them up for processing.
+	// The technique is valuable for rollup activities to avoid fanout and reduce resource contention.
+	// We could move this into a library if another component needed it.
+	// queue is indexed by namespace, so that we bundle up on a per-namespace basis
+	queue      *workqueue.Type
+	workLock   sync.Mutex
+	work       map[string][]*admissionWaiter
+	dirtyWork  map[string][]*admissionWaiter
+	inProgress sets.String
+
+	updatedQuotas     map[string]*quotaapi.ClusterResourceQuota
+	updatedQuotasLock sync.Mutex
+
+	lockFactory LockFactory
+}
+
+type admissionWaiter struct {
+	attributes admission.Attributes
+	finished   chan struct{}
+	result     error
+}
+
+type defaultDeny struct{}
+
+func (defaultDeny) Error() string {
+	return "DEFAULT DENY"
+}
+
+func IsDefaultDeny(err error) bool {
+	if err == nil {
+		return false
+	}
+
+	_, ok := err.(defaultDeny)
+	return ok
+}
+
+func newAdmissionWaiter(a admission.Attributes) *admissionWaiter {
+	return &admissionWaiter{
+		attributes: a,
+		finished:   make(chan struct{}),
+		result:     defaultDeny{},
+	}
+}
+
+// newQuotaEvaluator configures an admission controller that can enforce quota constraints
+// using the provided registry.  The registry must have the capability to handle group/kinds that
+// are persisted by the server this admission controller is intercepting
+func newQuotaEvaluator(client oclient.Interface, registry quota.Registry) (*quotaEvaluator, error) {
+	return &quotaEvaluator{
+		client:   client,
+		registry: registry,
+
+		queue:      workqueue.New(),
+		work:       map[string][]*admissionWaiter{},
+		dirtyWork:  map[string][]*admissionWaiter{},
+		inProgress: sets.String{},
+
+		lockFactory: NewDefaultLockFactory(),
+
+		updatedQuotas: map[string]*quotaapi.ClusterResourceQuota{},
+	}, nil
+}
+
+// Run begins watching and syncing.
+func (e *quotaEvaluator) Run(workers int) {
+	defer utilruntime.HandleCrash()
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(e.doWork, time.Second, make(chan struct{}))
+	}
+}
+
+func (e *quotaEvaluator) doWork() {
+	for {
```

Comment:
https://github.com/kubernetes/kubernetes/blob/master/pkg/util/wait/wait.go#L46

Seems like that plus what you have here means that the `wait.Util` call will not exit or jitter